cli/src/main/scala/com/lightbend/rp/reactivecli/runtime/kubernetes/Service.scala

```diff
@@ -54,44 +57,39 @@ object Service {
   def generate(annotations: Annotations,
                apiVersion: String,
                clusterIp: Option[String],
-               deploymentType: DeploymentType): Try[Option[Service]] =
-    if (annotations.endpoints.isEmpty)
-      Success(None)
-    else
-      (annotations.appName, annotations.version) match {
-        case (Some(rawAppName), Some(version)) =>
-          // FIXME there's a bit of code duplicate in Deployment
-          val appName = serviceName(rawAppName)
-          val appVersion = serviceName(s"$appName${Deployment.VersionSeparator}${version.version}")
+               deploymentType: DeploymentType): ValidationNel[String, Option[Service]] =
+    (annotations.appNameValidation |@| annotations.versionValidation) { (rawAppName, version) =>
+      // FIXME there's a bit of code duplicate in Deployment
+      val appName = serviceName(rawAppName)
+      val appVersion = serviceName(s"$appName${Deployment.VersionSeparator}${version.version}")
 
-          val selector =
-            deploymentType match {
-              case CanaryDeploymentType    => "app" -> appName.asJson
-              case RollingDeploymentType   => "app" -> appName.asJson
-              case BlueGreenDeploymentType => "appVersion" -> appVersion.asJson
-            }
-
-          Success(
-            Some(
-              Service(
-                appName,
-                Json(
-                  "apiVersion" -> apiVersion.asJson,
-                  "kind" -> "Service".asJson,
-                  "metadata" -> Json(
-                    "labels" -> Json(
-                      "app" -> appName.asJson),
-                    "name" -> appName.asJson)
-                    .deepmerge(
-                      annotations.namespace.fold(jEmptyObject)(ns => Json("namespace" -> serviceName(ns).asJson))),
-                  "spec" -> Json(
-                    "clusterIP" -> clusterIp.getOrElse("None").asJson,
-                    "ports" -> annotations.endpoints.asJson,
-                    "selector" -> Json(selector))))))
-        case _ =>
-          Failure(new IllegalArgumentException("Unable to generate Service resource: application name is required"))
-      }
+      val selector =
+        deploymentType match {
+          case CanaryDeploymentType    => "app" -> appName.asJson
+          case RollingDeploymentType   => "app" -> appName.asJson
+          case BlueGreenDeploymentType => "appVersion" -> appVersion.asJson
```

Comment:
Can I confirm for `BlueGreenDeploymentType` we are only checking the app version? This contains only the semver, i.e. `3.2.1` - so I do think we need the `appName` as well:

```
val selector =
        deploymentType match {
           case CanaryDeploymentType    => Json("app" -> appName.asJson)
           case RollingDeploymentType   => Json("app" -> appName.asJson)
           case BlueGreenDeploymentType => Json("app" -> appName.asJson, "appVersion" -> appVersion.asJson)
         }
```

Reply:
`appVersion` has `appName` in it .. it should probably be named `appNameVersion`.